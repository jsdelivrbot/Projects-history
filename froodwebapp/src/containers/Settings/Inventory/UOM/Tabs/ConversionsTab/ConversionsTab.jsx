import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Row, Col } from 'react-flexbox-grid';
import {
  uomConversionsGetRequest,
  uomConversionsSaveRequest,
  uomConversionsUpdateRequest,
  uomConversionsDeleteRequest
} from 'redux-base/actions';
import { Table, Spin, Select } from 'antd';
import { ConnectedTopFormModal, ActionButton } from 'components';
import { row, table, select } from 'styles/common.scss';
import columns from './conversionTabHelper';
import fields from './modalFields';


const mapStateToProps = state => ({
  uom: state.uom.data,
  conversions: state.uom.conversions,
  loadingPage: state.uom.loadingPage,
  needReloadUOMConversions: state.uom.needReloadUOMConversions,
});

const mapDispatchToProps = {
  uomConversionsGetRequest,
  uomConversionsSaveRequest,
  uomConversionsUpdateRequest,
  uomConversionsDeleteRequest
};

const newConversion = {
  id: '',
  fromUnit: '',
  fromQty: '',
  toId: '',
  toQty: '',
};

export class ConversionsTab extends Component {
  constructor(props) {
    super(props);
    this.state = {
      modalVisible: false,
      modalData: newConversion,
      selectedUOMId: ''
    };
  }

  componentWillReceiveProps(nextProps) {
    // when we delete of update conversion
    if (nextProps.needReloadUOMConversions) {
      this.setState({
        modalVisible: false,
        modalData: undefined
      }, () => this.props.uomConversionsGetRequest({
        id: this.state.selectedUOMId
      }));
      return;
    }

    // when we create conversion
    if (this.props.conversions.length !== nextProps.conversions.length) {
      this.setState({
        modalVisible: false,
        modalData: newConversion
      });
    }
  }

  handleToggleModal = () => {
    const { name } = this.props.uom.find(item => item.id === Number(this.state.selectedUOMId));
    this.setState({
      modalVisible: !this.state.modalVisible,
      modalData: !this.state.modalVisible ? { fromUnit: name } : newConversion
    });
  }

  handleSave = (data) => {
    if (this.state.modalData.toId) {
      this.props.uomConversionsUpdateRequest({
        id: this.state.selectedUOMId,
        payload: {
          fromQty: data.fromQty,
          id: data.id,
          toQty: data.toQty,
          toId: data.toId,
        }
      });
    } else {
      this.props.uomConversionsSaveRequest({
        id: this.state.selectedUOMId,
        payload: {
          fromQty: data.fromQty,
          toQty: data.toQty,
          toId: data.toId,
        }
      });
    }
  }

  handleDeactivate = (e) => {
    this.props.uomConversionsDeleteRequest({
      id: this.state.selectedUOMId,
      mappingId: Number(e.target.id),
    });
  }

  handleUOMSelect = (uomId) => {
    this.setState({
      selectedUOMId: uomId
    }, () => {
      this.props.uomConversionsGetRequest({
        id: uomId
      });
    });
  }

  handleEdit = (e) => {
    this.setState({
      modalVisible: true,
      modalData: this.props.conversions.find(item => item.id === Number(e.target.id))
    });
  }

  render() {
    const {
      modalData,
      modalVisible,
      selectedUOMId
    } = this.state;

    const {
      uom,
      conversions,
      loadingPage
    } = this.props;

    return (
      <div>
        <ConnectedTopFormModal
          loading={ loadingPage }
          title="New Conversion"
          buttonVisible={ false }
          handleSave={ this.handleSave }
          handleToggleModal={ this.handleToggleModal }
          visible={ modalVisible }
          initialValues={ modalData }
          fields={ fields(uom) }
        />
        <Row between="xs">
          <Col xs md={ 6 } lg={ 4 }>
            Conversion for UOM
            <Select
              className={ select }
              onChange={ this.handleUOMSelect }
            >
              { uom.map(item => (
                <Select.Option
                  key={ item.id }
                  value={ toString(item.id) }
                >
                  { item.name }
                </Select.Option>
              ))}
            </Select>
          </Col>
          <Col xs={ 1 } >
            <ActionButton
              style={ { float: 'right' } }
              disabled={ !selectedUOMId }
              onClick={ this.handleToggleModal }
            >
              New
            </ActionButton>
          </Col>
        </Row>
        <Spin spinning={ loadingPage }>
          <Row className={ row }>
            <Table
              className={ table }
              rowKey="id"
              small="small"
              columns={ columns(this.handleEdit, this.handleDeactivate) }
              dataSource={ conversions }
            />
          </Row>
        </Spin>
      </div>
    );
  }
}

ConversionsTab.propTypes = {
  // trigger
  loadingPage: PropTypes.bool.isRequired,
  needReloadUOMConversions: PropTypes.bool.isRequired,
  // data
  uom: PropTypes.array.isRequired,
  conversions: PropTypes.array.isRequired,
  // redux-base
  uomConversionsGetRequest: PropTypes.func.isRequired,
  uomConversionsSaveRequest: PropTypes.func.isRequired,
  uomConversionsUpdateRequest: PropTypes.func.isRequired,
  uomConversionsDeleteRequest: PropTypes.func.isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(ConversionsTab);
