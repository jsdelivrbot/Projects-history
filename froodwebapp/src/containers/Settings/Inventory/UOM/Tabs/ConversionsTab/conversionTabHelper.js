import React from 'react';
import { TableActionButtons } from 'components';

export default (handleEdit, handleDeactivate) => [{
  title: 'QUANTITY',
  dataIndex: 'fromQty'
}, {
  title: 'UOM',
  dataIndex: 'fromName'
}, {
  title: 'EQUALS',
  dataIndex: 'eq',
  render: () => '='
}, {
  title: 'QUANTITY',
  dataIndex: 'toQty'
}, {
  title: 'UOM',
  dataIndex: 'toName'
}, {
  title: 'ACTIONS',
  render: (text, record) => (
    <TableActionButtons
      record={ record }
      handleEdit={ handleEdit }
      handleDeactivate={ handleDeactivate }
      deleteButtonVisible
    />
  )
}];
