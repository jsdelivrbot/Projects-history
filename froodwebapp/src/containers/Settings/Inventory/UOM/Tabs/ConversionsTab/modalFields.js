import { getMenuItems } from 'utils';

export default uom => ([{
  header: 'Rate',
  type: 'text',
  name: 'fromQty',
  editableForExistingItem: true,
  editableForNewItem: true
}, {
  header: 'Unit',
  type: 'text',
  name: 'fromUnit',
  editableForExistingItem: false,
  editableForNewItem: true
}, {
  header: 'Rate',
  type: 'text',
  name: 'toQty',
  editableForExistingItem: true,
  editableForNewItem: true
}, {
  header: 'Unit',
  type: 'select',
  name: 'toId',
  editableForExistingItem: true,
  editableForNewItem: true,
  menuItems: getMenuItems(uom)
}]);
