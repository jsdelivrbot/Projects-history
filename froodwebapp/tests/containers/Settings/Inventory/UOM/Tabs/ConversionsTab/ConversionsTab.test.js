import { ConversionsTab } from 'containers/Settings/Inventory/UOM/Tabs/ConversionsTab/ConversionsTab';
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

const { describe, expect, it } = global;

describe('ConversionsTab', () => {
  const props = {
    loadingPage: false,
    needReloadUOMConversions: false,
    uom: [],
    conversions: [],
    uomConversionsGetRequest: jest.fn(),
    uomConversionsSaveRequest: jest.fn(),
    uomConversionsUpdateRequest: jest.fn(),
    uomConversionsDeleteRequest: jest.fn()
  };

  it('renders the ConversionsTab component', () => {
    const component = shallow(<ConversionsTab { ...props } />);

    expect(component.length).toEqual(1);
  });

  it('capturing Snapshot of ConversionsTab', () => {
    const component = shallow(<ConversionsTab { ...props } />);
    const view = toJson(component);
    const actual = expect(view);

    actual.toMatchSnapshot();
  });

  it('calls uomConversionsSaveRequest in handleSave with correct param', () => {
    const data = {
      fromQty: 10,
      toQty: 15,
      toId: 1
    };
    const selectedUOMId = '1';
    const component = shallow(<ConversionsTab { ...props } />);
    component.setState({
      selectedUOMId
    });
    component.instance().handleSave(data);

    expect(props.uomConversionsSaveRequest.mock.calls.length).toBe(1);

    expect(props.uomConversionsSaveRequest.mock.calls[0][0]).toEqual({
      id: selectedUOMId,
      payload: {
        fromQty: data.fromQty,
        toQty: data.toQty,
        toId: data.toId,
      }
    });
  });

  it('calls uomConversionsDeleteRequest in handleDeactivate with correct param', () => {
    const e = {
      target: {
        id: '1'
      }
    };
    const selectedUOMId = '1';
    const component = shallow(<ConversionsTab { ...props } />);
    component.setState({
      selectedUOMId
    });
    component.instance().handleDeactivate(e);

    expect(props.uomConversionsDeleteRequest.mock.calls.length).toBe(1);

    expect(props.uomConversionsDeleteRequest.mock.calls[0][0]).toEqual({
      id: selectedUOMId,
      mappingId: Number(e.target.id),
    });
  });

  it('calls uomConversionsDeleteRequest in handleUOMSelect with correct param', () => {
    const uomId = '1';
    const component = shallow(<ConversionsTab { ...props } />);
    component.setState({
      selectedUOMId: uomId
    });
    component.instance().handleUOMSelect(uomId);

    expect(props.uomConversionsGetRequest.mock.calls.length).toBe(1);

    expect(props.uomConversionsGetRequest.mock.calls[0][0]).toEqual({
      id: uomId
    });
  });
});
