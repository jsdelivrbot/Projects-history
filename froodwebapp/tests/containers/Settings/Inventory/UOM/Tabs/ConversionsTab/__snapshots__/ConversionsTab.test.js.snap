// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConversionsTab capturing Snapshot of ConversionsTab 1`] = `
<div>
  <ReduxForm
    buttonVisible={false}
    fields={
      Array [
        Object {
          "editableForExistingItem": true,
          "header": "Rate",
          "name": "fromQty",
          "type": "text",
        },
        Object {
          "editableForExistingItem": false,
          "header": "Unit",
          "name": "fromUnit",
          "type": "text",
        },
        Object {
          "editableForExistingItem": true,
          "header": "Rate",
          "name": "toQty",
          "type": "text",
        },
        Object {
          "editableForExistingItem": true,
          "header": "Unit",
          "menuItems": Array [],
          "name": "toId",
          "type": "select",
        },
      ]
    }
    handleSave={[Function]}
    handleToggleModal={[Function]}
    initialValues={
      Object {
        "fromQty": "",
        "fromUnit": "",
        "id": "",
        "toId": "",
        "toQty": "",
      }
    }
    loading={false}
    title="New Conversion"
    visible={false}
  />
  <Row
    between="xs"
  >
    <Col
      lg={4}
      md={6}
      xs={true}
    >
      Conversion for UOM
      <Select
        choiceTransitionName="zoom"
        className="select"
        onChange={[Function]}
        prefixCls="ant-select"
        showSearch={false}
        transitionName="slide-up"
      />
    </Col>
    <Col
      xs={1}
    >
      <ActionButton
        disabled={true}
        onClick={[Function]}
        style={
          Object {
            "float": "right",
          }
        }
      >
        New
      </ActionButton>
    </Col>
  </Row>
  <Spin
    prefixCls="ant-spin"
    size="default"
    spinning={false}
    wrapperClassName=""
  >
    <Row
      className="row"
    >
      <Table
        bordered={false}
        className="table"
        columns={
          Array [
            Object {
              "dataIndex": "fromQty",
              "title": "QUANTITY",
            },
            Object {
              "dataIndex": "fromName",
              "title": "UOM",
            },
            Object {
              "dataIndex": "eq",
              "render": [Function],
              "title": "EQUALS",
            },
            Object {
              "dataIndex": "toQty",
              "title": "QUANTITY",
            },
            Object {
              "dataIndex": "toName",
              "title": "UOM",
            },
            Object {
              "render": [Function],
              "title": "ACTIONS",
            },
          ]
        }
        dataSource={Array []}
        indentSize={20}
        loading={false}
        locale={Object {}}
        prefixCls="ant-table"
        rowKey="id"
        rowSelection={null}
        showHeader={true}
        size="large"
        small="small"
        useFixedHeader={false}
      />
    </Row>
  </Spin>
</div>
`;
